### ISSUE TITLE
Entry with data binding misses keystrokes or panics on fast input

### ISSUE THREAD

**Issue by AlexNigl** [fyne-io/fyne#3887]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

I'm using a Barcode-Scanner to input data into an "widget.NewEntryWithData". If the Data is long, and has about a 12ms delay between keystrokes, the input either gets cut or fyne panics.

The length of the string that is needed to hang the application seems to depend on the performance of the computer.
On slower machines I have seen this bug with string length of 10.

A normal Entry (without data binding) works as expected.

### How to reproduce

1. run example code
2. focus input
3. use Barcode scanner (or simulate it with something like xdotool)
4. Panic

```
panic: runtime error: slice bounds out of range [6:5]

goroutine 15 [running]:
fyne.io/fyne/v2/widget.(*RichText).insertAt(0xc0001942c0, 0x6, {0xc000535f20, 0x1})
        /home/alex/go/pkg/mod/fyne.io/fyne/v2@v2.3.4/widget/richtext.go:236 +0x390
fyne.io/fyne/v2/widget.(*Entry).TypedRune(0xc0001ba1c0, 0x72)
        /home/alex/go/pkg/mod/fyne.io/fyne/v2@v2.3.4/widget/entry.go:747 +0x1a5
fyne.io/fyne/v2/internal/driver/glfw.(*window).processCharInput.func1()
        /home/alex/go/pkg/mod/fyne.io/fyne/v2@v2.3.4/internal/driver/glfw/window.go:786 +0x25
fyne.io/fyne/v2/internal/driver/common.(*Window).RunEventQueue(0x0?)
        /home/alex/go/pkg/mod/fyne.io/fyne/v2@v2.3.4/internal/driver/common/window.go:35 +0x3e
created by fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).createWindow.func1
        /home/alex/go/pkg/mod/fyne.io/fyne/v2@v2.3.4/internal/driver/glfw/window.go:946 +0x136
exit status 2
```

### Example code

```go
package main

import (
        "fyne.io/fyne/v2/app"
        "fyne.io/fyne/v2/container"
        "fyne.io/fyne/v2/data/binding"
        "fyne.io/fyne/v2/widget"
)

func main() {
        myApp := app.New()
        window := myApp.NewWindow("Test")
        sb := binding.NewString()
        c := container.NewVBox(
                widget.NewEntryWithData(sb),
                widget.NewLabelWithData(sb),
        )
        window.SetContent(
                c,
        )
        window.ShowAndRun()
}
```

### Fyne version

2.3.4 and 2.3.5

### Go compiler version

1.20.3 and 1.20.5

### Operating system and version

Linux (6.1), Windows 10 and Windows 11

### Additional Information

On Linux (with X11) you can simulate fast input with:
```sh
go run . & sleep 5; xdotool type "thisisareallylongstringthatmostlikelykillsyourappandorwontappearinitsentirety0123456789"
```

**Comment by andydotxyz** [comment]
Making progress on this, it turns out there are 2 or 3 intertwining issues!

**Comment by andydotxyz** [comment]
Fixed on develop and release/v2.3.x branch

**Comment by williambrode** [comment]
@andydotxyz an unintended consequence of this change is that people intentionally setting the text twice in succession causes the text not to update.  I have something like this:

```
entry.SetText("")
entry.Bind(stringBinding)
```

And my text is empty everywhere (unless the binding fires).

But doesn't ignoring updates in the 100ms following an update also cause string bindings to potentially fail.  My binding is bound to process output, if it writes two lines within 100ms at the end can't my last line just never show up?
